modelcluster/fields.py

```diff
@@ -317,6 +322,46 @@ def get_queryset(self):
 
             return FakeQuerySet(rel_model, results)
 
+        def get_prefetch_queryset(self, instances, queryset=None):
+            # Derived from Django's ManyRelatedManager.get_prefetch_queryset.
+            if queryset is None:
+                queryset = super().get_queryset()
+
+            queryset._add_hints(instance=instances[0])
+            queryset = queryset.using(queryset._db or self._db)
+
+            query = {'%s__in' % query_field_name: instances}
+            queryset = queryset._next_is_sticky().filter(**query)
+
+            fk = self.through._meta.get_field(source_field_name)
+            join_table = fk.model._meta.db_table
+
+            connection = connections[queryset.db]
+            qn = connection.ops.quote_name
+
+            queryset = queryset.extra(select={
+                '_prefetch_related_val_%s' % f.attname:
+                '%s.%s' % (qn(join_table), qn(f.column)) for f in fk.local_related_fields})
+
+            return (
+                queryset,
+                lambda result: tuple(
+                    getattr(result, '_prefetch_related_val_%s' % f.attname)
+                    for f in fk.local_related_fields
+                ),
+                lambda inst: tuple(
+                    f.get_db_prep_value(getattr(inst, f.attname), connection)
+                    for f in fk.foreign_related_fields
+                ),
+                False,
+                relation_name,
+                False,
+            )
+
+        def _apply_rel_filters(self, queryset):
+            # Required for get_prefetch_queryset.
+            return queryset._next_is_sticky()
```

Comment:
@chosak This method doesn't appear to be needed any longer

Reply:
@ababic thanks for taking a look at this PR.

The [Django 1.10 release notes](https://docs.djangoproject.com/en/3.0/releases/1.10/#custom-manager-classes-available-through-prefetch-related-must-define-a-apply-rel-filters-method) say that

> If you defined a custom manager class available through `prefetch_related()` you must make sure it defines a `_apply_rel_filters()` method.

Prior to Django 2.0 this wasn't technically required yet, but 2.0 [removed a shim that allowed that](https://docs.djangoproject.com/en/3.0/releases/2.0/#features-removed-in-2-0):

> The shim for supporting custom related manager classes without a _apply_rel_filters() method is removed.

`_apply_rel_filters` gets called by Django code [here](https://github.com/django/django/blob/a4e6030904df63b3f10aa0729b86dc6942b0458e/django/db/models/query.py#L1885). Note that the DeferringRelatedManager used for ParentalKey also [defines this method](https://github.com/wagtail/django-modelcluster/blob/400c5422ad3069e71154b38ea21a9221aa8fb3ba/modelcluster/fields.py#L69).

Reply:
Ah, okay. That makes sense! In that case, ignore me! This is super impressive BTW... I've had a crack at this before and it's not the nicest part of Django to delve into.